url24d793285d_auzll_gengweibo_tgz-pJ8-org_json_TestJ8/src/com/gengweibo/dao/mysql/WeiboDaoJdbcImpl.java

```diff
@@ -0,0 +1,275 @@
+        if (null != stmt) {
+            try {
+                stmt.close();
+            } catch (SQLException e) {
+                if (LOG.isDebugEnabled()) {
+                    LOG.debug("method:closeQuietly", e);
+                }
+            }
+        }
+    }
+
+    private void closeQuietly(ResultSet rs) {
+        if (null != rs) {
+            try {
+                rs.close();
+            } catch (SQLException e) {
+                if (LOG.isDebugEnabled()) {
+                    LOG.debug("method:closeQuietly", e);
+                }
+            }
+        }
+    }
+
+    public void save(IWeibo weibo) {
+        Connection conn = null;
+        PreparedStatement pstmtQuery = null;
+        ResultSet rs = null;
+        PreparedStatement pstmtUpdate = null;
+        PreparedStatement pstmtSave = null;
+
+        try {
+            conn = dataSource.getConnection();
+            conn.setAutoCommit(false);
+
+            Weibo wei = (Weibo) weibo;
+
+            // 先查找，如果已存在就更新，否则直接新插入
+            pstmtQuery = conn
+                    .prepareStatement("select * from gwb_weibo where weibo_id=?");
+            pstmtQuery.setString(1, wei.getWeiboId());
+            rs = pstmtQuery.executeQuery();
+
+            if (rs.next()) {
+
+                // 更新
+                pstmtUpdate = conn
+                        .prepareStatement("update gwb_weibo set access_token=?,token_secret=?,account_id=? where weibo_id=?");
+                pstmtUpdate.setString(1, wei.getAccessor().accessToken);
+                pstmtUpdate.setString(2, wei.getAccessor().tokenSecret);
+                pstmtUpdate.setString(3, wei.getAccountId());
+                pstmtUpdate.setString(4, wei.getWeiboId());
+                pstmtUpdate.executeUpdate();
+
+                weibo.setSynUpdate(Boolean.valueOf(rs.getString("syn_update")));
+
+            } else {
+                // 新保存
+                pstmtSave = conn
+                        .prepareStatement("insert into gwb_weibo(account_id,access_token,token_secret,type,weibo_id,syn_update) values(?,?,?,?,?,?)");
+                pstmtSave.setString(1, wei.getAccountId());
+                pstmtSave.setString(2, wei.getAccessor().accessToken);
+                pstmtSave.setString(3, wei.getAccessor().tokenSecret);
+                pstmtSave.setString(4, wei.getType().getEnName());
+                pstmtSave.setString(5, wei.getWeiboId());
+
+                //
+                wei.setSynUpdate(true);
+                pstmtSave.setString(6, String.valueOf(wei.isSynUpdate()));
+
+                pstmtSave.executeUpdate();
+            }
+
+            conn.commit();
+
+        } catch (SQLException e) {
+            if (null != conn) {
+                try {
+                    conn.rollback();
+                } catch (SQLException re) {
+                    if (LOG.isDebugEnabled()) {
+                        LOG.debug("method:save", re);
+                    }
+                }
+            }
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 125:133, 161:169, 258:266.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.